VocalTractLabApi.cpp

```diff
@@ -93,88 +94,26 @@ BOOL APIENTRY DllMain( HANDLE hModule, DWORD  ul_reason_for_call, LPVOID lpReser
-  // Load the XML data from the speaker file.
-  // ****************************************************************
-
-  vector<XmlError> xmlErrors;
-  XmlNode *rootNode = xmlParseFile(string(speakerFileName), "speaker", &xmlErrors);
-  if (rootNode == NULL)
-  {
-    xmlPrintErrors(xmlErrors);
-    return false;
-  }
-
-  // ****************************************************************
-  // Load the data for the glottis models.
-  // ****************************************************************
-
-  // This may be overwritten later.
-  selectedGlottis = GEOMETRIC_GLOTTIS;
-
-  XmlNode *glottisModelsNode = rootNode->getChildElement("glottis_models");
-  if (glottisModelsNode != NULL)
-  {
-    int i;
-    XmlNode *glottisNode;
-
-    for (i=0; (i < (int)glottisModelsNode->childElement.size()) && (i < NUM_GLOTTIS_MODELS); i++)
+    //// ****************************************************************
+	//// Load the speaker from XML file
+	//// ****************************************************************
+    Speaker speaker(speakerFileName);
+
+    *vocalTract = *speaker.getVocalTract();
+    const auto glottisInfo = speaker.getGlottisModels();
+    int i = 0;
+	for (const auto& g : glottisInfo)
     {
-      glottisNode = glottisModelsNode->childElement[i];
-      if (glottisNode->getAttributeString("type") == glottis[i]->getName())
-      {
-        if (glottisNode->getAttributeInt("selected") == 1)
-        {
-          selectedGlottis = i;
-        }
-        if (glottis[i]->readFromXml(*glottisNode) == false)
-        {
-          printf("Error: Failed to read glottis data for glottis model %d!\n", i);
-          delete rootNode;
-          return false;
-        }
-      }
-      else
-      {
-        printf("Error: The type of the glottis model %d in the speaker file is '%s' "
-          "but should be '%s'!\n", i, 
-          glottisNode->getAttributeString("type").c_str(), 
-          glottis[i]->getName().c_str());
-
-        delete rootNode;
-        return false;
-      }
-    }
-  }
-  else
-  {
-    printf("Warning: No glottis model data found in the speaker file %s!\n", speakerFileName);
-  }
+        glottis[i++] = g;
+	}
+    selectedGlottis = speaker.getSelectedGlottis();
 
-  // Free the memory of the XML tree !
-  delete rootNode;
-
-  // ****************************************************************
-  // Load the vocal tract anatomy and vocal tract shapes.
-  // ****************************************************************
-
-  try
-  {
-    vocalTract->readFromXml(string(speakerFileName));
-    vocalTract->calculateAll();
-  }
-  catch (std::string st)
-  {
-    printf("%s\n", st.c_str());
-    printf("Error reading the anatomy data from %s.\n", speakerFileName);
-    return false;
-  }
 
   return true;
 }
 
 
+
 // ****************************************************************************
 // Init. the synthesis with the given speaker file name, e.g. "JD2.speaker".
```

Comment:
JD3.speaker
You may also search the code for the term JD2, as it should be replaced everywhere.

Reply:
Done